compiler/code-gen/files/type-tagger.cpp

```diff
@@ -41,51 +47,95 @@ void TypeTagger::compile(CodeGenerator &W) const {
     "array< class_instance<C$VK$TL$RpcResponse> >"
   };
 
-  IncludesCollector collector;
-
-  for (auto type : forkable_types) {
-    sorted_types.insert(type_out(type, gen_out_style::tagger));
-    collector.add_all_class_types(*type);
-  }
-  for (auto type : waitable_types) {
-    collector.add_all_class_types(*type);
+  for (const auto *type : forkable_types_) {
+    sorted_types.emplace(type_out(type, gen_out_style::tagger));
   }
 
   std::map<int, std::string> hashes;
 
-  W << collector << NL;
-
-  for (const std::string &type : sorted_types) {
-    int hash = vk::std_hash(type);
+  for (const auto &type : sorted_types) {
+    const auto hash = static_cast<int>(vk::std_hash(type));
     kphp_assert(hash);
-    kphp_assert(hashes.insert({hash, type}).second);
+    kphp_assert(hashes.emplace(hash, type).second);
+  }
+
+  return hashes;
+}
+
+void TypeTagger::compile_tagger(CodeGenerator &W, const IncludesCollector &includes, const std::map<int, std::string> &hash_of_types) const noexcept {
+  W << OpenFile{"_tagger.cpp"};
+  W << ExternInclude{G->settings().runtime_headers.get()};
+  W << includes << NL;
+
+  for (const auto &[hash, type] : hash_of_types) {
     W << "template<>" << NL;
-    FunctionSignatureGenerator(W) << "int Storage::tagger<" << type << ">::get_tag() " << BEGIN;
+    FunctionSignatureGenerator{W} << "int Storage::tagger<" << type << ">::get_tag() " << BEGIN;
     W << "return " << hash << ";" << NL;
     W << END << NL << NL;
   }
 
+  W << CloseFile{};
+}
+
+void TypeTagger::compile_loader_header(CodeGenerator &W, const IncludesCollector &includes, const std::map<int, std::string> &hash_of_types) const noexcept {
+  W << OpenFile{loader_file_};
```

Comment:
Here you use `_loader_file`, whereas above and before "_tagger.cpp" explicitly, without any constant. I'd suggest to use "_loader.h" directly here also.

BTW, what do you think of renaming `_loader.h` to `_tagger_loader.h` and renaming `_loader_instantiations{i}.cpp` to `_tagger_instantiations{i}.cpp`?

Reply:
> Here you use _loader_file, whereas above and before "_tagger.cpp" explicitly, without any constant. I'd suggest to use "_loader.h" directly here also.


This is on purpose. "_loader.h" constant used in 2 places so to keep consistency I used variable. "_tagger.cpp" used in one place, so just KISS here.


> BTW, what do you think of renaming _loader.h to _tagger_loader.h and renaming _loader_instantiations{i}.cpp to _tagger_instantiations{i}.cpp?

`
  template<typename T>
  struct tagger {
    static int get_tag() noexcept;
  };

  template<typename T>
  struct loader {
    static loader_fun get_function(int tag) noexcept;
  };`
Struct loader and tagger has no coupling actually, so I think better to leave as is.